### PR TITLE
Add typehints for torchvision.io

### PR DESCRIPTION
**Pull request by pmeier**
I'll do this first before finishing `torchvision.datasets` since it is needed for the video datasets.

### CODE REVIEW

File: torchvision/io/_video_opt.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
 
-import importlib
+import importlib.machinery
```
Comment (pmeier) [inline torchvision/io/_video_opt.py:2]: According to `mypy`, the second argument to `importlib.machinery.FileFinder` should be `Tuple[Loader, List[str]]` (see [CI fail](https://app.circleci.com/pipelines/github/pytorch/vision/3448/workflows/01102881-92e1-4d9e-addb-f6869f8d80f6/jobs/193510)). Thus it expects a `Loader` instance, but we are currently passing a class (see L18-23). Is this intentional? If so, I'll just add an ignore.

File: torchvision/io/_video_opt.py [hunk 2]
```diff
@@ -306,13 +306,13 @@ def _read_video_from_memory(
     video_height=0,  # type: int
```
Comment (pmeier) [inline torchvision/io/_video_opt.py:306]: Can I switch this to Python 3 type annotations or will this mess with `torch.script`?

File: torchvision/io/_video_opt.py [hunk 3]
```diff
@@ -306,13 +306,13 @@ def _read_video_from_memory(
...
     video_min_dimension=0,  # type: int
     video_max_dimension=0,  # type: int
-    video_pts_range=(0, -1),  # type: List[int]
+    video_pts_range=None,  # type: Optional[List[int]]
```
Comment (pmeier) [inline torchvision/io/_video_opt.py:309]: This was probably done to avoid a mutable default argument. Are we fixed on `List`s and `Tuple`s or can we use `Sequence`? If not, I don't see a better way than mine to also pass type checks.

Comment (fmassa) [reply torchvision/io/_video_opt.py:309]: We are fixed with `List` because we need this to work with torchscript.

File: torchvision/io/_video_opt.py (hunk 2 above)
Comment (fmassa) [reply torchvision/io/_video_opt.py:306]: we can use Python3 type annotations now

File: torchvision/io/image.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ def decode_png(input):
     Returns:
         output (Tensor[image_width, image_height, 3])
     """
-    if not isinstance(input, torch.Tensor) or input.numel() == 0 or input.ndim != 1:
+    if not isinstance(input, torch.Tensor) or input.numel() == 0 or input.dim() != 1:
```
Comment (fmassa) [inline torchvision/io/image.py:40]: any reason to change `ndim` to `dim()`? `ndim` is a new attribute of tensor, which follows numpy more closely.

File: torchvision/io/_video_opt.py [hunk 5]
```diff
@@ -378,9 +378,15 @@ def _read_video_from_memory(
...
-    _validate_pts(audio_pts_range)
+    if video_pts_range is None:
+        video_pts_range = [0, -1]
+    else:
+        _validate_pts(video_pts_range)
+
+    if audio_pts_range is None:
+        audio_pts_range = [0, -1]
+    else:
+        _validate_pts(audio_pts_range)
```
Comment (fmassa) [inline torchvision/io/_video_opt.py:389]: Let's keep the old behavior more consistent by always calling `_validate_pts`
```suggestion
    if video_pts_range is None:
        video_pts_range = [0, -1]

    if audio_pts_range is None:
        audio_pts_range = [0, -1]

    _validate_pts(video_pts_range)
    _validate_pts(audio_pts_range)
```

File: torchvision/io/_video_opt.py (hunk 3 above)
Comment (fmassa) [reply torchvision/io/_video_opt.py:309]: The way you are doing it is good

File: torchvision/io/_video_opt.py (hunk 1 above)
Comment (fmassa) [reply torchvision/io/_video_opt.py:2]: Hum, interesting. `FileFinder` seems to support both [an instance (callable) and a class](https://docs.python.org/3/library/importlib.html#importlib.util.LazyLoader.factory).

I would not want to change this as of now without further extensive testing, so I would go for an ignore and a comment describing this for now.

File: torchvision/io/image.py (hunk 4 above)
Comment (pmeier) [reply torchvision/io/image.py:40]: `mypy` doesn't recognize the `ndim` attribute:

```python
import torch

tensor = torch.empty(1, 2, 3)

print(tensor.dim())
print(tensor.ndim)
```

```
test_ndim.py:6: error: "Tensor" has no attribute "ndim"  [attr-defined]
    print(tensor.ndim)
          ^
Found 1 error in 1 file (checked 1 source file)
```

Either we use `dim()` or I add an ignore. If we later use [`warn_unused_ignores`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-warn-unused-ignores) we can simply remove the ignore as soon as this is fixed upstream.

File: torchvision/io/video.py [hunk 6]
```diff
@@ -287,15 +302,15 @@ def _can_read_timestamps_from_packets(container):
...
+def _decode_video_timestamps(container: av.container.Container) -> List[int]:
     if _can_read_timestamps_from_packets(container):
         # fast path
         return [x.pts for x in container.demux(video=0) if x.pts is not None]
     else:
         return [x.pts for x in container.decode(video=0) if x.pts is not None]
 
 
-def read_video_timestamps(filename, pts_unit="pts"):
+def read_video_timestamps(filename: str, pts_unit: str = "pts") -> Tuple[List[int], int]:
```
Comment (pmeier) [inline torchvision/io/video.py:313]: This docstring states that `video_fps` should be `int` where as the docstring of [`_read_video_from_file`](https://github.com/pytorch/vision/blob/df6a7960920534d53d31f6d06fa89dea90cc1b50/torchvision/io/_video_opt.py#L216-L218) states that it should be `float`. Which one is correct?

File: torchvision/io/video.py [hunk 7]
```diff
@@ -260,16 +275,16 @@ def read_video(filename, start_pts=0, end_pts=None, pts_unit="pts"):
...
+        vframes = torch.as_tensor(np.stack(vframes_list))
     else:
         vframes = torch.empty((0, 1, 1, 3), dtype=torch.uint8)
 
-    if aframes:
-        aframes = np.concatenate(aframes, 1)
+    if aframes_list:
+        aframes = np.concatenate(aframes_list, 1)
         aframes = torch.as_tensor(aframes)
         aframes = _align_audio_frames(aframes, audio_frames, start_pts, end_pts)
```
Comment (pmeier) [inline torchvision/io/video.py:289]: `_align_audio_frames` uses `start_pts` and `end_pts` as indices. Here `end_pts` by default is `float("inf")`, which will not work.

**Comment by pmeier** [comment]
Binary CI failures are related. I'll address them first and ping you if they are fixed.

File: torchvision/io/image.py (hunk 4 above)
Comment (fmassa) [reply torchvision/io/image.py:38]: Can you open an issue in PyTorch mentioning the `ndim` doesn't have mypy support?

File: torchvision/io/_video_opt.py [hunk 8]
```diff
@@ -416,7 +429,9 @@ def _read_video_from_memory(
     return vframes, aframes
 
 
-def _read_video_timestamps_from_memory(video_data):
+def _read_video_timestamps_from_memory(
+    video_data: Union[torch.Tensor, BufferedIOBase]
```
Comment (fmassa) [inline torchvision/io/_video_opt.py:433]: we should not forget that those type signatures are also used for torchscript, and torchscript doesn't support Union, and I'm not sure about `BufferedIOBase`, so let's just leave `Tensor` here.

File: torchvision/io/video.py (hunk 7 above)
Comment (fmassa) [reply torchvision/io/video.py:289]: At least in the implementation, we have `end > end_ref`, and given that `end_ref` is `inf`, this branch will never hit.
But do you mean that the type annotations will never work here?

File: torchvision/io/_video_opt.py [hunk 9]
```diff
@@ -470,7 +484,9 @@ def _probe_video_from_memory(video_data):
     return info
 
 
-def _read_video(filename, start_pts=0, end_pts=None, pts_unit="pts"):
+def _read_video(
+    filename: str, start_pts: int = 0, end_pts: Optional[float] = None, pts_unit: str = "pts"
```
Comment (fmassa) [inline torchvision/io/_video_opt.py:488]: we are in a complicated situation here because the type of `start_pts` and `end_pts` depend on `pts_unit`.

They are both floats if `pts_unit == 'sec'`, and both int if `pts_unit == 'pts'`. But I would rather not make this a `Union` here because it will hinder support for torchscript, specially in this file which is currently used in some production systems (with tests not necessarily showing up in GitHub's CI).

File: torchvision/io/video.py [hunk 10]
```diff
@@ -179,7 +192,9 @@ def _align_audio_frames(aframes, audio_frames, ref_start, ref_end):
     return aframes[:, s_idx:e_idx]
 
 
-def read_video(filename, start_pts=0, end_pts=None, pts_unit="pts"):
+def read_video(
+    filename: str, start_pts: int = 0, end_pts: Optional[float] = None, pts_unit: str = "pts"
```
Comment (fmassa) [inline torchvision/io/video.py:196]: just make `end_pts` be an optional of float here for now I think

**Review by fmassa — COMMENTED** [review]
Sorry for the delay in reviewing @pmeier 

Given that we are lacking CI testing on GitHub for `_video_opt` for now, can you please revert the changes to that file, disable mypy for it for now, and then maybe in the future once we have FFmpeg being compiled and tested in the CI we re-add it back?

File: torchvision/io/image.py (hunk 4 above)
Comment (pmeier) [reply torchvision/io/image.py:38]: Already reported and fixed https://github.com/pytorch/pytorch/issues/42908

Comment (pmeier) [reply torchvision/io/image.py:38]: Do you want to go with `dim()` or `ndim` and an ignore?

File: torchvision/io/video.py (hunk 7 above)
Comment (pmeier) [reply torchvision/io/video.py:289]: Correct:

```python
foo = [1, 2, 3]
foo[1:3.0]
```

```
error: Slice index must be an integer or None  [misc]
```

File: torchvision/io/_video_opt.py (hunk 9 above)
Comment (pmeier) [reply torchvision/io/_video_opt.py:488]: We do not need a `Union` here, since `int` is a subtype of `float` for `mypy`

```python
foo: int = 1.0
bar: float = 1
```

```
error: Incompatible types in assignment (expression has type "float", variable has type "int")  [assignment]
```

This is only a problem if we use `start_pts` or `end_pts` as indexes. But then, I need to take another look why we are trying to do float indexing in the first place.

File: torchvision/io/video.py (hunk 10 above)
Comment (pmeier) [reply torchvision/io/video.py:196]: Do you mean `start_pts`? `end_pts` is already an `Optional[float]`.

**Comment by pmeier** [comment]
> Sorry for the delay in reviewing @pmeier
> 
> Given that we are lacking CI testing on GitHub for `_video_opt` for now, can you please revert the changes to that file, disable mypy for it for now, and then maybe in the future once we have FFmpeg being compiled and tested in the CI we re-add it back?

Sure, I can do that. My motivation to go for `torchvision.io` stems from the fact that `torchvision.datasets.video_utils` makes heavy use of it. This includes `_video_opt`. Thus, if I remove those type hints again, we probably need careful manual checking for video datasets, since `mypy` can't help there.

File: torchvision/io/image.py (hunk 4 above)
Comment (fmassa) [reply torchvision/io/image.py:38]: If it's been fixed, can we keep `ndim`?

File: torchvision/io/video.py (hunk 7 above)
Comment (fmassa) [reply torchvision/io/video.py:289]: Can we just make the type annotations be `Optional[int]`?

File: torchvision/io/_video_opt.py (hunk 9 above)
Comment (fmassa) [reply torchvision/io/_video_opt.py:488]: We don't do float indexing. We just use `inf` as a way of saying: go until the end of the file. We could have used a different representation, like `-1` for this, but it would have made a few things a bit more complicated I think

File: torchvision/io/video.py (hunk 10 above)
Comment (fmassa) [reply torchvision/io/video.py:196]: Sorry, I meant `Optional[int]`

**Comment by fmassa** [comment]
Yeah, I would prefer if we leave `_video_opt` untouched for now. We can leave the video datasets last wrt typing.

File: torchvision/io/video.py (hunk 6 above)
Comment (pmeier) [reply torchvision/io/video.py:313]: @fmassa This is the last piece before `mypy` is happy.

File: torchvision/io/video.py [hunk 11]
```diff
@@ -89,8 +95,13 @@ def write_video(filename, video_array, fps: Union[int, float], video_codec="libx
 
 
 def _read_from_stream(
-    container, start_offset, end_offset, pts_unit, stream, stream_name
-):
+    container: "av.container.Container",
+    start_offset: int,
+    end_offset: float,
```
Comment (fmassa) [inline torchvision/io/video.py:100]: Given that `pts_unit` can also be `sec`, meaning that `start_offset` is a float, I think it might be preferable to have `start_offset` being a `float` as well. According to https://stackoverflow.com/questions/50928592/mypy-type-hint-unionfloat-int-is-there-a-number-type
> when an argument is annotated as having type float, an argument of type int is acceptable;

But this is not blocking for this PR, and can be done in a follow-up PR

File: torchvision/io/video.py (hunk 6 above)
Comment (fmassa) [reply torchvision/io/video.py:313]: We do
```python
video_fps = float(video_stream.average_rate)
```
so we return a float.

**Review by fmassa — COMMENTED** [review]
Last couple of comments and then I think we can merge this!

**Comment by pmeier** [comment]
@fmassa Test failures seem unrelated.

**Review by fmassa — APPROVED** [review]
Thanks!
